### PR TITLE
feat: Make DSPY instrumentation read from attributes from context

### PR DESCRIPTION
**Pull request by fjcasti1**

### CODE REVIEW

File: python/instrumentation/openinference-instrumentation-dspy/pyproject.toml [hunk 1]
```diff
@@ -27,6 +27,7 @@ dependencies = [
   "opentelemetry-api",
   "opentelemetry-instrumentation",
   "opentelemetry-semantic-conventions",
+  "openinference-instrumentation>=0.1.2",
```
Comment (axiomofjoy) [inline python/instrumentation/openinference-instrumentation-dspy/pyproject.toml:30]: This is the yanked version.

File: python/instrumentation/openinference-instrumentation-dspy/examples/dspy_basic_qa.py [hunk 2]
```diff
@@ -1,22 +1,23 @@
 import os
 
 import dspy
+from openinference.instrumentation import using_attributes
 from openinference.instrumentation.dspy import DSPyInstrumentor
 from opentelemetry import trace as trace_api
-from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
+
+# from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
```
Comment (axiomofjoy) [inline python/instrumentation/openinference-instrumentation-dspy/examples/dspy_basic_qa.py:8]: Can we add back in the OTLP exporter?

File: python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py [hunk 3]
```diff
@@ -221,6 +222,7 @@ def __call__(
                 )
             ),
         ) as span:
+            span.set_attributes(dict(get_attributes_from_context()))
```
Comment (axiomofjoy) [inline python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py:225]:
```
        with self._tracer.start_as_current_span(
            span_name,
            attributes={
                **_flatten(
                    {
                        OPENINFERENCE_SPAN_KIND: LLM.value,
                        LLM_MODEL_NAME: instance.kwargs.get("model"),
                        LLM_INVOCATION_PARAMETERS: json.dumps(invocation_parameters),
                        INPUT_VALUE: str(prompt),
                        INPUT_MIME_TYPE: OpenInferenceMimeTypeValues.TEXT.value,
                    }
                ),
                **get_attributes_from_context(),
            },
        ) as span:
```

non-blocking suggestion

File: python/instrumentation/openinference-instrumentation-dspy/tests/openinference/instrumentation/dspy/test_instrumentor.py [hunk 4]
```diff
@@ -417,19 +564,449 @@ def exact_match(example: dspy.Example, pred: dspy.Example, trace: Any = None) ->
...
+    assert isinstance(chain_attributes.pop(OUTPUT_VALUE), str)
+    assert (
+        OpenInferenceMimeTypeValues(chain_attributes.pop(OUTPUT_MIME_TYPE))
+        == OpenInferenceMimeTypeValues.JSON
+    )
+    _check_context_attributes(chain_attributes, session_id, user_id, metadata, tags)
+    assert chain_attributes == {}
+
+
+def test_google_lm_with_context_attributes(
```
Comment (axiomofjoy) [inline python/instrumentation/openinference-instrumentation-dspy/tests/openinference/instrumentation/dspy/test_instrumentor.py:676]: Do we need this test if we have the OpenAI LM test for `using_attributes`?

**Review by axiomofjoy — APPROVED** [review]
awesome, just need to increment the pinned `openinference-instrumentation` version

File: python/instrumentation/openinference-instrumentation-dspy/examples/dspy_basic_qa.py (hunk 2 above)
Comment (fjcasti1) [reply python/instrumentation/openinference-instrumentation-dspy/examples/dspy_basic_qa.py:8]: Left over from test, my bad

File: python/instrumentation/openinference-instrumentation-dspy/tests/openinference/instrumentation/dspy/test_instrumentor.py (hunk 4 above)
Comment (fjcasti1) [reply python/instrumentation/openinference-instrumentation-dspy/tests/openinference/instrumentation/dspy/test_instrumentor.py:676]: Didn't we want to test with and without?

File: python/instrumentation/openinference-instrumentation-dspy/examples/dspy_basic_qa.py (hunk 2 above)
Comment (fjcasti1) [reply python/instrumentation/openinference-instrumentation-dspy/examples/dspy_basic_qa.py]: Done

File: python/instrumentation/openinference-instrumentation-dspy/pyproject.toml (hunk 1 above)
Comment (fjcasti1) [reply python/instrumentation/openinference-instrumentation-dspy/pyproject.toml]: Done

File: python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py (hunk 3 above)
Comment (fjcasti1) [reply python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py:225]: mypy panics and I did not care enough to force my way into it :(

**Pull request merged**